File: ui/src/store/insights.ts
import { create } from "zustand";
import type { SetState } from "zustand";
import type { Insight } from "../api";

interface InsightsState {
  insights: Insight[];
  modalInsight: Insight | null;
  setInsights: (items: Insight[]) => void;
  addInsight: (item: Insight) => void;
  removeInsight: (id: string) => void;
  updateInsightTitle: (id: string, title: string) => void;
  updateInsightText: (id: string, text: string) => void;
  setModalInsight: (ins: Insight | null) => void;
}

export const useInsightsStore = create<InsightsState>((set: SetState<InsightsState>) => ({
  insights: [],
  modalInsight: null,
  setInsights: (items: Insight[]) => set({ insights: items }),
  addInsight: (item: Insight) => set((state) => ({ insights: [item, ...state.insights] })),
  removeInsight: (id: string) =>
    set((state) => ({ insights: state.insights.filter((ins) => ins.id !== id) })),
  updateInsightTitle: (id: string, title: string) =>
    set((state) => {
      const updatedInsights = state.insights.map((ins) => (ins.id === id ? { ...ins, title } : ins));
      const updatedModal = state.modalInsight && state.modalInsight.id === id ? { ...state.modalInsight, title } : state.modalInsight;
      return { insights: updatedInsights, modalInsight: updatedModal } as any;
    }),
  updateInsightText: (id: string, text: string) =>
    set((state) => {
      const updatedInsights = state.insights.map((ins) => (ins.id === id ? { ...ins, text } : ins));
      const updatedModal = state.modalInsight && state.modalInsight.id === id ? { ...state.modalInsight, text } : state.modalInsight;
      return { insights: updatedInsights, modalInsight: updatedModal } as any;
    }),
  setModalInsight: (ins: Insight | null) => set({ modalInsight: ins }),
}) as InsightsState as any); 

File: native-ui/src/state/insights.rs
use serde::{Serialize, Deserialize};
use crate::api::models::Insight;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct InsightsState {
    pub insights: Vec<Insight>,
    pub modal_insight: Option<Insight>,
}

impl InsightsState {
    pub fn new() -> Self {
        Self {
            insights: Vec::new(),
            modal_insight: None,
        }
    }

    pub fn set_insights(&mut self, insights: Vec<Insight>) {
        self.insights = insights;
    }

    pub fn add_insight(&mut self, insight: Insight) {
        self.insights.insert(0, insight);
    }

    pub fn remove_insight(&mut self, id: &str) {
        self.insights.retain(|i| i.id != id);
        if let Some(ref modal) = self.modal_insight {
            if modal.id == id {
                self.modal_insight = None;
            }
        }
    }

    pub fn update_insight_title(&mut self, id: &str, title: String) {
        for insight in &mut self.insights {
            if insight.id == id {
                insight.title = title.clone();
            }
        }
        if let Some(ref mut modal) = self.modal_insight {
            if modal.id == id {
                modal.title = title;
            }
        }
    }

    pub fn update_insight_text(&mut self, id: &str, text: String) {
        for insight in &mut self.insights {
            if insight.id == id {
                insight.text = text.clone();
            }
        }
        if let Some(ref mut modal) = self.modal_insight {
            if modal.id == id {
                modal.text = text;
            }
        }
    }

    pub fn set_modal_insight(&mut self, insight: Option<Insight>) {
        self.modal_insight = insight;
    }
}

impl Default for InsightsState {
    fn default() -> Self {
        Self::new()
    }
}

